packages/components/src/internal/util/RouteLeave.tsx

```diff
@@ -1,96 +1,135 @@
+ * @param confirmMessage: The confirm message you want to display to the user, this message is only displayed when
+ * navigating away from the page, not when closing the tab or browser window. Browsers do not let you customize the
+ * message displayed when the browser/tab is closed.
  */
-export const withRouteLeave = (Component: React.ComponentType) => {
-    return class RouteLeaveHOCImpl extends React.Component<RouteLeaveProps> {
-        componentDidMount(): void {
-            // attach the hook to the current route, which will be the last index of the routes prop
-            this.props.router.setRouteLeaveHook(this.props.routes[this.props.routes.length - 1], this.onRouteLeave);
+export const useRouteLeave = (
+    router: InjectedRouter,
+    routes: PlainRoute[],
+    confirmMessage = CONFIRM_MESSAGE
+): GetSetIsDirty => {
+    const initialHistoryLength = useMemo(() => history.length, []);
+    const isDirty = useRef<boolean>(false);
+
+    const setIsDirty = useCallback(
+        (dirty: boolean) => {
+            isDirty.current = dirty;
+        },
+        [isDirty]
+    );
+
+    const getIsDirty = useCallback((): boolean => {
+        return isDirty.current;
+    }, [isDirty]);
+
+    const onRouteLeave = useCallback(() => {
+        if (isDirty.current === true) {
+            const result = confirm(confirmMessage);
+
+            if (!result) {
+                // Issue 42101: the browser changes the URL before onRouteLeave is called, so if the user cancels
+                // we need to go back to the URL we were just at. Unfortunately going back to the URL we were just at
+                // depends on what the user did, and detecting that is spotty at best. If the user clicked a link we
+                // need to go back in history, if the user hit the back button we need to go forward in history. The
+                // "best" way to detect this is by keeping track of the history length when we first render. If the user
+                // clicked a link this will increase. This is method is unfortunately not perfect, but it prevents us
+                // from showing the cancel dialog for every entry in the user's history until they either reach the
+                // beginning of history, or hit leave page.
+                //
+                // You can still end up in a bad state in two known scenarios:
+                // - if you alternate between using the back button, hitting cancel, clicking a link, hitting cancel
+                // - If your browser history is at maximum length, this appears to be 50 in Chrome
+                //
+                // To avoid this entire situation we need to either commit to using Link components everywhere, which
+                // don't cause the URL to change before onRouteLeave is called, or we need to upgrade to a newer version
+                // of React Router.
+                if (history.length > initialHistoryLength) {
```

Comment:
Per our discussion, we should make note of the nicety that the history length is capped (at 50 in Chrome for me), so this will not always do the right thing.   We need to document that at least.

The previous implementation didn't rely on this length.  Could a similar approach be used here? Is the "event" available to be passed as an argument?

Reply:
Added. The previous implementation needed the event because it was also handling unload, by separating out unload and routeLeave events we don't need that extra logic to check for POP.